README.md
# Flutter Boom Menu

## Usage

The BoomMenu widget is built to be placed in the `Scaffold.floatingActionButton` argument, replacing the `FloatingActionButton` widget.
It's not possible to set its position with the `Scaffold.floatingActionButtonLocation` argument, but it's possible to set right/bottom margin with the `marginRight` and `marginBottom` arguments (default to 16) to place the button anywhere in the screen.
Using the `Scaffold.bottomNavigationBar` the floating button will be always placed above the bar, so the `BottomAppBar.hasNotch` should be always `false`.

**Title**

Every child button can have a `Icon`,`Title`, `SubTitle` which can be customized providing by yourself. If the `Title` parameter is not provided the title will be not rendered.

The package will handle the animation by itself.

![alt text](https://github.com/AimenSayoud/flutter_boom_menu_new/screenshot/boom-menu-video.gif)

**Example Usage ( complete with all params ):**

```dart
Widget build(BuildContext context) {
    return Scaffold(
        floatingActionButton: BoomMenu(
      animatedIcon: AnimatedIcons.menu_close,
      animatedIconTheme: IconThemeData(size: 22.0),
      //child: Icon(Icons.add),
      onOpen: () => print('OPENING DIAL'),
      onClose: () => print('DIAL CLOSED'),
      scrollVisible: scrollVisible,
      overlayColor: Colors.black,
      overlayOpacity: 0.7,
      children: [
        MenuItem(
          child: Icon(Icons.accessibility, color: Colors.black),
          title: "Profiles",
          titleColor: Colors.white,
          subtitle: "You Can View the Noel Profile",
          subTitleColor: Colors.white,
          backgroundColor: Colors.deepOrange,
          onTap: () => print('FIRST CHILD'),
        ),
        MenuItem(
          child: Icon(Icons.brush, color: Colors.black),
          title: "Profiles",
          titleColor: Colors.white,
          subtitle: "You Can View the Noel Profile",
          subTitleColor: Colors.white,
          backgroundColor: Colors.green,
          onTap: () => print('SECOND CHILD'),
        ),
        MenuItem(
          child: Icon(Icons.keyboard_voice, color: Colors.black),
          title: "Profile",
          titleColor: Colors.white,
          subtitle: "You Can View the Noel Profile",
          subTitleColor: Colors.white,
          backgroundColor: Colors.blue,
          onTap: () => print('THIRD CHILD'),
        ),
        MenuItem(
          child: Icon(Icons.ac_unit, color: Colors.black),
          title: "Profiles",
          titleColor: Colors.white,
          subtitle: "You Can View the Noel Profile",
          subTitleColor: Colors.white,
          backgroundColor: Colors.blue,
          onTap: () => print('FOURTH CHILD'),
        )
      ],
    ),
    );
}
```

## Issues & Feedback

Please file an [issue](https://github.com/AimenSayoud/flutter_boom_menu_new/issues) to send feedback or report a bug. Thank you!

## Contributing

Every pull request is welcome.


CHANGELOG.md
## 1.0.0

* add the null-safety and migrate the code to flutter 3


lib/src/boom_menu.dart
import 'package:flutter/material.dart';

import 'animated_child.dart';
import 'animated_floating_button.dart';
import 'background_overlay.dart';
import 'boom_menu_item.dart';

/// Builds the Speed Dial
class BoomMenu extends StatefulWidget {
  /// Children buttons, from the lowest to the highest.
  final List<MenuItemModel> children;

  /// Used to get the button hidden on scroll. See examples for more info.
  final bool scrollVisible;

  final Color? backgroundColor;
  final Color? foregroundColor;
  final double elevation;
  final ShapeBorder fabMenuBorder;
  final Alignment fabAlignment;

  final double marginLeft;
  final double marginRight;
  final double marginBottom;

  final double fabPaddingLeft;
  final double fabPaddingRight;
  final double fabPaddingTop;

  /// The color of the background overlay.
  final Color overlayColor;

  /// The opacity of the background overlay when the dial is open.
  final double overlayOpacity;

  /// The animated icon to show as the main button child. If this is provided the [child] is ignored.
  final AnimatedIconData? animatedIcon;

  /// The theme for the animated icon.
  final IconThemeData? animatedIconTheme;

  /// The child of the main button, ignored if [animatedIcon] is non [null].
  final Widget? child;

  /// Executed when the dial is opened.
  final VoidCallback? onOpen;

  /// Executed when the dial is closed.
  final VoidCallback? onClose;

  /// Executed when the dial is pressed. If given, the dial only opens on long press!
  final VoidCallback? onPress;

  /// If true user is forced to close dial manually by tapping main button. WARNING: If true, overlay is not rendered.
  final bool overlayVisible;

  /// The speed of the animation
  final int animationSpeed;

  final String? title;
  final String? subtitle;
  final Color? titleColor;
  final Color? subTitleColor;

  BoomMenu(
      {this.children = const [],
      this.scrollVisible = true,
      this.title,
      this.subtitle,
      this.backgroundColor,
      this.titleColor,
      this.subTitleColor,
      this.foregroundColor,
      this.elevation = 6.0,
      this.overlayOpacity = 0.8,
      this.overlayColor = Colors.white,
      this.animatedIcon,
      this.animatedIconTheme,
      this.child,
      this.marginBottom = 0,
      this.marginLeft = 16,
      this.marginRight = 0,
      this.onOpen,
      this.onClose,
      this.overlayVisible = false,
      this.fabMenuBorder = const CircleBorder(),
      this.fabAlignment = Alignment.centerRight,
      this.fabPaddingRight = 0,
      this.fabPaddingLeft = 0,
      this.fabPaddingTop = 0,
      this.onPress,
      this.animationSpeed = 150});

  @override
  _BoomMenuState createState() => _BoomMenuState();
}

class _BoomMenuState extends State<BoomMenu>
    with SingleTickerProviderStateMixin {
  AnimationController? _controller;

  bool _open = false;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: _calculateMainControllerDuration(),
      vsync: this,
    );
  }

  Duration _calculateMainControllerDuration() => Duration(
      milliseconds: widget.animationSpeed +
          widget.children.length * (widget.animationSpeed / 5).round());

  @override
  void dispose() {
    _controller!.dispose();
    super.dispose();
  }

  void _performAnimation() {
    if (!mounted) return;
    if (_open) {
      _controller!.forward();
    } else {
      _controller!.reverse();
    }
  }

  @override
  void didUpdateWidget(BoomMenu oldWidget) {
    if (oldWidget.children.length != widget.children.length) {
      _controller!.duration = _calculateMainControllerDuration();
    }

    super.didUpdateWidget(oldWidget);
  }

  void _toggleChildren() {
    var newValue = !_open;
    setState(() {
      _open = newValue;
    });
    if (newValue && widget.onOpen != null) {
      widget.onOpen!();
    }
    _performAnimation();
    if (!newValue && widget.onClose != null) {
      widget.onClose!();
    }
  }

  List<Widget> _getChildrenList() {
    final singleChildrenTween = 1.0 / widget.children.length;

    return widget.children
        .map((MenuItemModel child) {
          int index = widget.children.indexOf(child);

          var childAnimation = Tween(begin: 0.0, end: 62.0).animate(
            CurvedAnimation(
              parent: this._controller!,
              curve: Interval(
                0,
                singleChildrenTween * (index + 1),
              ),
            ),
          );

          return AnimatedChild(
              animation: childAnimation,
              index: index,
              visible: _open,
              backgroundColor: child.backgroundColor!,
              elevation: child.elevation!,
              child: child.child!,
              title: child.title!,
              subtitle: child.subtitle!,
              titleColor: child.titleColor!,
              subTitleColor: child.subTitleColor!,
              onTap: child.onTap!,
              toggleChildren: () {
                if (!widget.overlayVisible) _toggleChildren();
              });
        })
        .toList()
        .reversed
        .toList();
  }

  Widget _renderOverlay() {
    return Positioned(
      right: -16.0,
      bottom: -16.0,
      top: _open ? 0.0 : null,
      left: _open ? 0.0 : null,
      child: GestureDetector(
        onTap: _toggleChildren,
        child: BackgroundOverlay(
          animation: _controller!,
          color: widget.overlayColor,
          opacity: widget.overlayOpacity,
        ),
      ),
    );
  }

  Widget _renderButton() {
    var child = widget.animatedIcon != null
        ? AnimatedIcon(
            icon: widget.animatedIcon!,
            progress: _controller!,
            color: widget.animatedIconTheme!.color,
            size: widget.animatedIconTheme!.size,
          )
        : widget.child;

    var fabChildren = _getChildrenList();

    var animatedFloatingButton = AnimatedFloatingButton(
        visible: widget.scrollVisible,
        backgroundColor: widget.backgroundColor,
        foregroundColor: widget.foregroundColor,
        elevation: widget.elevation,
        onLongPress: _toggleChildren,
        callback: (_open || widget.onPress == null)
            ? _toggleChildren
            : widget.onPress,
        child: child,
        shape: widget.fabMenuBorder);

    return Positioned(
      left: widget.marginLeft + 16,
      bottom: widget.marginBottom,
      right: widget.marginRight,
      child: Container(
        alignment: Alignment.bottomCenter,
        width: MediaQuery.of(context).size.width,
        height: MediaQuery.of(context).size.height,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.end,
          crossAxisAlignment: CrossAxisAlignment.end,
          /*children: List.from(fabChildren)
            ..add(
              Container(
                margin: EdgeInsets.only(top: 8.0, right: 2.0),
                child: animatedFloatingButton,
              ),
            ),*/
          children: <Widget>[
            SizedBox(height: kToolbarHeight + 40),
            Visibility(
              visible: _open,
              child: Expanded(
                child: ListView(
                  children: List.from(fabChildren),
                  reverse: true,
                ),
              ),
            ),
            Align(
              alignment: widget.fabAlignment,
              child: Container(
                padding: EdgeInsets.only(
                  left: widget.fabPaddingLeft,
                  right: widget.fabPaddingRight,
                  top: 8.0 + widget.fabPaddingTop,
                ),
                child: animatedFloatingButton,
              ),
            ),
          ],
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final children = [
      !widget.overlayVisible ? _renderOverlay() : Container(),
      _renderButton(),
    ];

    return Stack(
      alignment: Alignment.bottomRight,
      fit: StackFit.expand,
      clipBehavior: Clip.none,
      children: children,
    );
  }
}


lib/src/animated_child.dart
import 'package:flutter/material.dart';

class AnimatedChild extends AnimatedWidget {
  final int index;
  final Color backgroundColor;
  final double elevation;
  final Widget child;

  final bool visible;
  final VoidCallback onTap;
  final VoidCallback toggleChildren;
  final String title;
  final String subtitle;
  final Color titleColor;
  final Color subTitleColor;

  AnimatedChild(
      {Key? key,
      Animation<double>? animation,
      required this.index,
      required this.backgroundColor,
      this.elevation = 6.0,
      required this.child,
      required this.title,
      required this.subtitle,
      this.visible = false,
      required this.onTap,
      required this.toggleChildren,
      required this.titleColor,
      required this.subTitleColor})
      : super(key: key, listenable: animation!);

  void _performAction() {
    if (onTap != null) onTap();
    toggleChildren();
  }

  Widget build(BuildContext context) {
    final Animation<double> animation = listenable as Animation<double>;

    final Widget buttonChild = animation.value > 50.0
        ? Container(
            width: animation.value,
            height: animation.value,
            child: Row(
              children: <Widget>[
                Padding(
                  padding: const EdgeInsets.all(10.0),
                  child: child ?? Container(),
                ),
                Expanded(
                  child: Padding(
                    padding: const EdgeInsets.only(left: 8.0, right: 8.0),
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: <Widget>[
                        Text(
                          title,
                          overflow: TextOverflow.ellipsis,
                          style: TextStyle(
                              color: (titleColor == null)
                                  ? Colors.black
                                  : titleColor,
                              fontSize: 16.0),
                        ),
                        SizedBox(height: 3.0),
                        Text(
                          subtitle,
                          overflow: TextOverflow.ellipsis,
                          style: TextStyle(
                              color: (subTitleColor == null)
                                  ? Colors.black
                                  : subTitleColor,
                              fontSize: 12.0),
                        )
                      ],
                    ),
                  ),
                )
              ],
            ),
          )
        : Container(
            width: 0.0,
            height: 0.0,
          );

    return Container(
      width: MediaQuery.of(context).size.width - 30,
      height: 80.0,
      padding: EdgeInsets.only(bottom: 72 - animation.value),
      child: GestureDetector(
        onTap: _performAction,
        child: Card(
          elevation: 2,
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(7.0),
          ),
          //width: animation.value,
          color: backgroundColor,
          child: buttonChild,
        ),
      ),
    );
  }
}


lib/src/animated_floating_button.dart
import 'package:flutter/material.dart';

class AnimatedFloatingButton extends StatelessWidget {
  final bool visible;
  final VoidCallback? callback;
  final VoidCallback onLongPress;
  final Widget? child;
  final Color? backgroundColor;
  final Color? foregroundColor;
  final String? tooltip;
  final String? heroTag;
  final double elevation;
  final ShapeBorder shape;
  final Curve curve;

  AnimatedFloatingButton({
    this.visible = true,
    this.callback,
    this.child,
    this.backgroundColor,
    this.foregroundColor,
    this.tooltip,
    this.heroTag,
    this.elevation = 6.0,
    this.shape = const CircleBorder(),
    this.curve = Curves.linear,
    required this.onLongPress,
  });

  @override
  Widget build(BuildContext context) {
    var margin = visible ? 0.0 : 28.0;

    return Container(
      constraints: const BoxConstraints(
        minHeight: 0.0,
        minWidth: 0.0,
      ),
      width: 56.0,
      height: 56.0,
      child: AnimatedContainer(
        curve: curve,
        margin: EdgeInsets.all(margin),
        duration: Duration(milliseconds: 150),
        width: visible ? 56.0 : 0.0,
        height: visible ? 56.0 : 0.0,
        child: GestureDetector(
          onLongPress: onLongPress,
          child: FloatingActionButton(
            child: visible ? child : null,
            backgroundColor: backgroundColor,
            foregroundColor: foregroundColor,
            onPressed: callback,
            tooltip: tooltip,
            heroTag: heroTag,
            elevation: elevation,
            highlightElevation: elevation,
            shape: shape,
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
/// Created by aimen 2022-07-27.
/// Email : aymensayoud2019@gmail.com

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

import 'package:flutter_boom_menu_new/flutter_boom_menu_new.dart';

void main() {
  runApp(MaterialApp(home: MyApp(), title: 'Boom Menu Example'));
}

class MyApp extends StatefulWidget {
  @override
  MyAppState createState() => MyAppState();
}

class MyAppState extends State<MyApp> with TickerProviderStateMixin {
  ScrollController? scrollController;
  bool scrollVisible = true;

  @override
  void initState() {
    super.initState();

    scrollController = ScrollController()
      ..addListener(() {
        setDialVisible(scrollController!.position.userScrollDirection ==
            ScrollDirection.forward);
      });
  }

  void setDialVisible(bool value) {
    setState(() {
      scrollVisible = value;
    });
  }

  Widget buildBody() {
    return ListView.builder(
      controller: scrollController,
      itemCount: 30,
      itemBuilder: (ctx, i) => ListTile(title: Text('Item $i')),
    );
  }

  BoomMenu buildBoomMenu() {
    return BoomMenu(
        animatedIcon: AnimatedIcons.menu_close,
        animatedIconTheme: IconThemeData(size: 22.0),
        //child: Icon(Icons.add),
        onOpen: () => print('OPENING DIAL'),
        onClose: () => print('DIAL CLOSED'),
        scrollVisible: scrollVisible,
        overlayColor: Colors.black,
        elevation: 10,
        overlayOpacity: 0.7,
        children: [
          MenuItemModel(
//          child: Icon(Icons.accessibility, color: Colors.black, size: 40,),
            child:
                Image.asset('assets/logout_icon.png', color: Colors.grey[850]),
            title: "Logout",
            titleColor: Colors.grey[850]!,
            subtitle: "Lorem ipsum dolor sit amet, consectetur",
            subTitleColor: Colors.grey[850]!,
            backgroundColor: Colors.grey[50]!,
            onTap: () => print('THIRD CHILD'),
            elevation: 10,
          ),
          MenuItemModel(
            child: Image.asset('assets/schemes_icon.png', color: Colors.white),
            title: "List",
            titleColor: Colors.white,
            subtitle: "Lorem ipsum dolor sit amet, consectetur",
            subTitleColor: Colors.white,
            backgroundColor: Colors.pinkAccent,
            onTap: () => print('FOURTH CHILD'),
            elevation: 10,
          ),
          MenuItemModel(
            child: Image.asset('assets/customers_icon.png',
                color: Colors.grey[850]),
            title: "Team",
            titleColor: Colors.grey[850]!,
            subtitle: "Lorem ipsum dolor sit amet, consectetur",
            subTitleColor: Colors.grey[850]!,
            backgroundColor: Colors.grey[50]!,
            onTap: () => print('THIRD CHILD'),
            elevation: 10,
          ),
          MenuItemModel(
            child: Image.asset('assets/profile_icon.png', color: Colors.white),
            title: "Profile",
            titleColor: Colors.white,
            subtitle: "Lorem ipsum dolor sit amet, consectetur",
            subTitleColor: Colors.white,
            backgroundColor: Colors.blue,
            onTap: () => print('FOURTH CHILD'),
            elevation: 10,
          )
        ]);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Boom Menu Example')),
      body: buildBody(),
      floatingActionButton: buildBoomMenu(),
    );
  }
}
